src/syntax.rs

```diff
@@ -0,0 +1,105 @@
+pub enum DiscreteType {
+    Int,
+    Bool,
+    Float,
+}
+
+pub enum Type {
+    Vector(usize, DiscreteType),
+    Function,
+    Shader,
+}
+
+pub enum ShaderType {
+    Vector,
+    Fragment,
+    Tesselation,
+    // And more!
+}
+
+pub enum BOp {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    And,
+    Or,
+    Eq,
+    Neq,
+    Gt,
+    Gte,
+    Lt,
+    Lte,
+    BitAnd,
+    BitOr,
+    BitXor,
+}
+
+pub struct Block {
+    pub body: Vec<Box<Expr>>
+}
+pub struct Parameter {
+    name: String,
+    datatype: Type
+}
+
+pub enum AccessType {
+    Name(String),
+    Idx(usize),
+}
+
+pub enum Expr {
+    I(usize),
+    B(bool),
+    F(f32),
+    Vect {
+        is_matrix: bool,
+        datatype: DiscreteType,
+        value: Block,
+    },
+    Def {
+        mutable: bool,
+        name: String,
+        value: Box<Expr>,
+    },
+    Ref(String),
+    App {
+        fname: String,
+        arguments: Block,
```

Comment:
Shouldn't application args be a list of `Box<Expr>`? I.e. `Vec<Box<Expr>>`

Reply:
Oh right, gotcha didn't see that about `Block` the first time around.

In regards to whether it adds anything, I'm inclined to agree, but it's making me think about how a vector of exprs is really kind of a vector of statements, but not really :/....

Reply:
Weren't we taking an Imperative syntax? i.e. a list of statements with the last statement evaluating to the value of the block

Reply:
We are. It was more a comment on how this feels like we're mixing Statements and Expressions together here, when they're really separate. Expressions should always resolve to final values, but statements do not necessarily do that. For example, with the way this is written now, we can write an expression ```BinOp Add 1 (Update "x" 5)```, but that's contingent on Update being a valid Expr, which would need to resolve to a final value. We *could* make Update resolve to the value of the updated name, but I don't know about that...

Reply:
To clarify, in some languages statements can constitute valid expressions, like `x = y = 2`. Where the value of the statement is just the value that is assigned to a name. This kind of muddies the waters however, but it's an option we could take if we can justify a final value for all 'statements', such that they constitute valid expressions (and then we can keep everything under the category of Expr like it is now).